network/dag/tree/iblt.go

```diff
@@ -0,0 +1,295 @@
+				updated = true
+			}
+		}
+
+		// if no pures exist, the iblt is empty or cannot be decoded
+		if !updated {
+			for _, b := range i.Buckets {
+				if !b.isEmpty() {
+					return remaining, missing, ErrDecodeNotPossible
+				}
+			}
+			return remaining, missing, nil
+		}
+	}
+}
+
+func (i Iblt) numBuckets() int {
+	return len(i.Buckets)
+}
+
+func makeBuckets(numBuckets int) []*bucket {
+	buckets := make([]*bucket, numBuckets)
+	for i := 0; i < numBuckets; i++ {
+		buckets[i] = new(bucket)
+	}
+	return buckets
+}
+
+func (i Iblt) bucketIndices(hash uint64) []uint32 {
+	bucketUsed := make(map[uint32]bool, i.K)
+	var indices []uint32
+	hashKeyBytes, nextBytes := make([]byte, 8), make([]byte, 4)
+	binary.LittleEndian.PutUint64(hashKeyBytes, hash)
+	next := murmur3.SeedSum32(i.Hk, hashKeyBytes)
+	for len(indices) < int(i.K) {
+		bucketId := next % uint32(i.numBuckets())
+		if !bucketUsed[bucketId] {
+			indices = append(indices, bucketId)
+			bucketUsed[bucketId] = true
+		}
+		binary.LittleEndian.PutUint32(nextBytes, next)
+		next = murmur3.SeedSum32(i.Hk, nextBytes)
+	}
+	return indices
+}
+
+func (i Iblt) hashKey(key hash.SHA256Hash) uint64 {
+	return murmur3.SeedSum64(i.Hc, key.Slice())
+}
+
+func (i Iblt) MarshalBinary() ([]byte, error) {
+	data := make([]byte, i.numBuckets()*bucketBytes)
+	for idx, b := range i.Buckets {
+		bs, err := b.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		copy(data[idx*bucketBytes:], bs)
+	}
+	return data, nil
+}
+
+func (i *Iblt) UnmarshalBinary(data []byte) error {
+	numBuckets := len(data) / bucketBytes
+	if len(data) != numBuckets*bucketBytes {
+		return errors.New("invalid data length")
+	}
+	buf := bytes.NewBuffer(data)
+	i.Hc = ibltHc
+	i.Hk = ibltHk
+	i.K = ibltK
+	i.Buckets = makeBuckets(numBuckets)
+	for j := 0; j < i.numBuckets(); j++ {
+		err := i.Buckets[j].UnmarshalBinary(buf.Next(bucketBytes))
+		if err != nil {
+			return fmt.Errorf("unmarshalling failed - %w", err)
+		}
+	}
+	return nil
+}
+
+// bucket
+type bucket struct {
+	// Count is signed to allow for negative counts after subtraction
+	Count   int32           `json:"count"`    // #1
+	HashSum uint64          `json:"hash_sum"` // #2
+	KeySum  hash.SHA256Hash `json:"key_sum"`  // #3
+}
+
+func (b *bucket) insert(key hash.SHA256Hash, hash uint64) {
+	b.Count++
+	b.update(key, hash)
+}
+
+func (b *bucket) delete(key hash.SHA256Hash, hash uint64) {
+	b.Count--
+	b.update(key, hash)
+}
+
+func (b *bucket) add(o *bucket) {
+	b.Count += o.Count
+	b.update(o.KeySum, o.HashSum)
+}
+
+func (b *bucket) subtract(o *bucket) {
+	b.Count -= o.Count
+	b.update(o.KeySum, o.HashSum)
+}
+
+func (b *bucket) update(key hash.SHA256Hash, hash uint64) {
+	xor(&b.KeySum, b.KeySum, key)
+	b.HashSum ^= hash
+}
+
+func (b bucket) clone() *bucket {
+	return &b
+}
+
+func (b bucket) isEmpty() bool {
+	return b.equals(*new(bucket))
+}
+
+func (b bucket) equals(o bucket) bool {
+	return b.Count == o.Count && b.HashSum == o.HashSum && b.KeySum == o.KeySum
```

Comment:
I think you should be able to compare `bucket` by simly using the equality operation (`==`), because all fields are value types? In other workds, `b == o`.

Reply:
replaced with `b == o`